File: monitorlibrary/lib/data/project_position.dart
import 'package:geocoding/geocoding.dart';
import 'package:meta/meta.dart';
import 'package:monitorlibrary/data/position.dart' as ar;

import '../functions.dart';
import 'city.dart';

class ProjectPosition {
  String? projectName, projectId, caption, created, projectPositionId, organizationId;
  ar.Position? position;
  Placemark? placemark;
  List<City>? nearestCities;

  ProjectPosition(
      {required this.projectName,
      required this.caption,
        required this.projectPositionId,
      required this.created,
      required this.position,
      this.placemark,
      required this.nearestCities,
        required this.organizationId,
      required this.projectId});

  ProjectPosition.fromJson(Map data) {
    //pp(' 💜 ProjectPosition.fromJson: log 0');
    this.projectName = data['projectName'];
    //pp(' 💜 ProjectPosition.fromJson: log 1');
    this.projectId = data['projectId'];
    this.projectPositionId = data['projectPositionId'];
    //pp(' 💜 ProjectPosition.fromJson: log 2');
    this.caption = data['caption'];
    this.projectId = data['projectId'];
    this.organizationId = data['organizationId'];
    this.created = data['created'];
    //pp(' 💜 ProjectPosition.fromJson: log 3');

    if (data['position'] != null) {
      this.position = ar.Position.fromJson(data['position']);
    }
    //pp(' 💜 ProjectPosition.fromJson: log 4');
    if (data['placemark'] != null) {
      this.placemark = Placemark.fromMap(data['position']);
    }
    //pp(' 💜 ProjectPosition.fromJson: log 5');
    this.nearestCities = [];
    if (data['nearestCities'] != null) {
      List list = data['nearestCities'];
      list.forEach((c) {
        nearestCities!.add(City.fromJson(c));
      });
    }
    //pp(' 💜 ProjectPosition.fromJson: log end');
  }

  Map<String, dynamic> toJson() {
    var list = [];
    this.nearestCities!.forEach((c) {
      list.add(c.toJson());
    });
    Map<String, dynamic> map = {
      'projectName': projectName,
      'projectId': projectId,
      'organizationId': organizationId,
      'projectPositionId': projectPositionId,
      'caption': caption,
      'created': created,
      'position': position == null ? null : position!.toJson(),
      'placemark': placemark == null ? null : placemark!.toJson(),
      'nearestCities': list,
    };
    return map;
  }
}

File: backend/src/main/java/com/monitor/backend/data/ProjectPosition.java
package com.monitor.backend.data;

import org.springframework.data.mongodb.core.mapping.Document;

import java.util.List;

@Document(collection = "projectPosition")
public class ProjectPosition {

    private String projectId, projectPositionId;
    private Position position;
    private String projectName;
    private String caption;
    private String created;
    private Placemark placemark;
    private List<City> nearestCities;


    public ProjectPosition(String projectPositionId, String projectId, Position position, String projectName, String caption, String created, Placemark placemark, List<City> nearestCities) {
        this.projectId = projectId;
        this.projectPositionId = projectPositionId;
        this.position = position;
        this.projectName = projectName;
        this.caption = caption;
        this.created = created;
        this.placemark = placemark;
        this.nearestCities = nearestCities;
    }

    public ProjectPosition() {
    }

    public String getProjectPositionId() {
        return projectPositionId;
    }

    public void setProjectPositionId(String projectPositionId) {
        this.projectPositionId = projectPositionId;
    }

    public List<City> getNearestCities() {
        return nearestCities;
    }

    public void setNearestCities(List<City> nearestCities) {
        this.nearestCities = nearestCities;
    }

    public Placemark getPlacemark() {
        return placemark;
    }

    public void setPlacemark(Placemark placemark) {
        this.placemark = placemark;
    }

    public String getProjectId() {
        return projectId;
    }

    public void setProjectId(String projectId) {
        this.projectId = projectId;
    }

    public Position getPosition() {
        return position;
    }

    public void setPosition(Position position) {
        this.position = position;
    }

    public String getProjectName() {
        return projectName;
    }

    public void setProjectName(String projectName) {
        this.projectName = projectName;
    }

    public String getCaption() {
        return caption;
    }

    public void setCaption(String caption) {
        this.caption = caption;
    }

    public String getCreated() {
        return created;
    }

    public void setCreated(String created) {
        this.created = created;
    }
}
